lib/database/repo.ex

```diff
@@ -1,48 +1,55 @@
-defmodule BorsNG.Database.Repo do
+defmodule DynamicEctoRepoWrapper do
   @moduledoc """
-  An ecto data repository;
-  this process interacts with your persistent database.
-
-  Do not confuse this with a GitHub repo.
-  We call those `Project`s internally.
+  Exposes a macro to define Ecto.Repo functions dynamically.
   """
+  defmacro create_ecto_repo_callback(args, name) do
+    quote bind_quoted: [args: args, name: name] do
+      def unquote(name)(unquote_splicing(args)) do
+        repo = :persistent_term.get(:db_repo)
+        apply(repo, unquote(name), unquote(args))
+      end
+    end
+  end
 
-  use Ecto.Repo,
-    otp_app: :bors,
-    adapter: Application.get_env(:bors, BorsNG.Database.Repo)[:adapter]
+  def create_ecto_repo_callback_args(_, 0) do
+    []
+  end
 
-  def init(_, config) do
-    # Backwards compatibility hack: if POSTGRES_HOST is set, and the database URL is left at default,
-    # use the older configuration.
-    config = Confex.Resolver.resolve!(config)
-    no_host = is_nil(System.get_env("POSTGRES_HOST"))
+  def create_ecto_repo_callback_args(module, arity) do
+    Enum.map(1..arity, &Macro.var(:"arg#{&1}", module))
+  end
+end
 
-    config =
-      case config[:url] do
-        _ when no_host ->
-          config
+defmodule BorsNG.Database.Repo do
+  @moduledoc """
+  This is an Ecto.Repo wrapper that defines all callback functions.
+  """
+  import DynamicEctoRepoWrapper
 
-        "postgresql://postgres:Postgres1234@localhost/bors_dev" ->
-          [
-            adapter: Ecto.Adapters.Postgres,
-            username: "postgres",
-            password: "Postgres1234",
-            database: "bors_dev",
-            hostname: {:system, "POSTGRES_HOST", "localhost"},
-            pool_size: 10
-          ]
+  @ecto_repo_callbacks Path.join([__DIR__, "repo_callbacks.txt"])
+                       |> File.read!()
+                       |> String.trim()
+                       |> String.split("\n")
+                       |> Enum.map(fn line ->
+                         [line, arity] = line |> String.split("/") |> Enum.map(&String.trim(&1))
+                         {String.to_atom(line), String.to_integer(arity)}
+                       end)
 
-        "postgresql://postgres:Postgres1234@localhost/bors_test" ->
-          [
-            adapter: Ecto.Adapters.Postgres,
-            username: "postgres",
-            password: "Postgres1234",
-            database: "bors_test",
-            hostname: {:system, "POSTGRES_HOST", "localhost"},
-            pool_size: 10
-          ]
-      end
+  @ecto_repo_callbacks
+  |> Enum.flat_map(fn
+    {callback, arity} when arity >= 1 ->
+      is_defined = Enum.any?(@ecto_repo_callbacks, fn fun -> fun == {callback, arity - 1} end)
 
-    {:ok, config}
-  end
+      if is_defined,
+        do: [{callback, arity}],
+        else: [{callback, arity}, {callback, arity - 1}]
```

Comment:
we are not super excited about this way to define callbacks that take into account all arities. it would be much nicer if we could define the functions with default arguments in the first place, but we couldn't make that work with `Macro.var`.

Maybe you have another idea @notriddle?